### PR TITLE
fix: support pre-computation of constants in Step Functions

### LINKED ISSUE #83: Step Functions: negative numbers with UnaryExpr
A number, `-10` will be parsed as a `UnaryExpr(NumberLiteralExpr())`, but the Step Functions interpreter (and maybe the VTL interpreter) do not properly handle this case.

### PR DESCRIPTION
**Pull request by sam-goodwin**
Fixes #83 

Adds support for negative literal numbers in Step Functions.

```ts
const a = -1;
const a = 1 + 2;
const a = [-1];
task(-1);
list.slice(-1) // [-1:]
list.slice(0, -1); // [0:-1]
```

Support `+`, `-`, `/` and `*` for constants.

```ts
1 + 2
-1
1 + 2
"hello"
"hello" + " world"
"hello" + 1
1 + "hello"
"hello" + true
false + "hello"
null + "hello"
"hello" + null
"hello" + { place: "world" }
"hello" + ["world"];

const array = ["world"];
"hello" + array;

const object = { place: "world" };
"hello" + object;
```

Surface errors in IDE when computations are performed on dynamic values in a Step Function .

```ts
(input) => input.i + 1
//               ~~~~~~~~
```

### CODE REVIEW

File: test/step-function.test.ts [hunk 1]
```diff
@@ -190,9 +236,11 @@ test("let and set", () => {
     a = true;
     a = false;
     a = 0;
+    a = -1;
```
Comment (thantos) [inline test/step-function.test.ts:239]: do you support other negation?

`-(100)`

`- var`

Comment (sam-goodwin) [reply test/step-function.test.ts:239]: `-var` is not supported. Parens is a good catch.

Comment (sam-goodwin) [reply test/step-function.test.ts:239]: `-(100)` worked but i never handled a ParenthesisExpr, did we auto-remove them?

Comment (sam-goodwin) [reply test/step-function.test.ts:239]: I should also resolve constant computations, e.g. `1 + -1`

**Comment by sam-goodwin** [comment]
Also adding constant pre-evaluation support, eg adding numbers, concatenation of strings, etc

File: src/compile.ts [hunk 2]
```diff
@@ -1101,7 +1101,6 @@ export function compile(
                   const flattened = flattenDestructuredAssignment(
                     symbol.valueDeclaration
                   );
-                  console.log(ts.SyntaxKind[flattened.kind]);
```
Comment (thantos) [inline src/compile.ts:1104]: woops, thanks

File: src/expression.ts [hunk 3]
```diff
@@ -320,6 +322,7 @@ export type UnaryOp = "!" | "-";
 export class UnaryExpr extends BaseExpr<"UnaryExpr"> {
   constructor(readonly op: UnaryOp, readonly expr: Expr) {
     super("UnaryExpr");
+    // @ts-ignore
```
Comment (thantos) [inline src/expression.ts:325]: why?

File: src/validate.ts [hunk 4]
```diff
@@ -69,4 +46,50 @@ export function validate(
...
+        newError(
+          node,
+          100,
+          "Cannot perform arithmetic on variables in Step Function"
+        ),
+      ];
+    }
+    return [];
+  }
+  function newError(
```
Comment (thantos) [inline src/validate.ts:80]: nit: new line?

File: test/test-files/step-function.ts [hunk 5]
```diff
@@ -6,5 +6,30 @@ const app = new App({
 });
 const stack = new Stack(app, "stack");
 
-// arithmetic is not supported by Amazon States Language
-new StepFunction(stack, "F", () => 1 + 2);
+// unsupported arithmetic
```
Comment (thantos) [inline test/test-files/step-function.ts:9]: nit: plus objects. arrays, symbols, and functions?

**Review by thantos — CHANGES_REQUESTED** [review]

File: src/asl.ts [hunk 6]
```diff
@@ -1350,33 +1359,45 @@ export namespace ASL {
       return elementAccessExprToJsonPath(expr);
     }
 
+    debugger;
```
[the remark this answers is not in the document]
Comment (sam-goodwin) [reply src/asl.ts:1362]: Yep, leaving them in the production version of the code does not affect anyone. Noticed this pattern in the TS compiler.

File: src/asl.ts [hunk 7]
```diff
@@ -1256,11 +1270,8 @@ export namespace ASL {
       return toJsonPath(expr);
     } else if (expr.kind === "PropAccessExpr") {
       return `${toJson(expr.expr)}.${expr.name}`;
-    } else if (
-      expr.kind === "ElementAccessExpr" &&
-      expr.element.kind === "NumberLiteralExpr"
-    ) {
-      return `${toJson(expr.expr)}[${expr.element.value}]`;
+    } else if (expr.kind === "ElementAccessExpr") {
+      return toJsonPath(expr);
```
[the remark this answers is not in the document]
Comment (sam-goodwin) [reply src/asl.ts:1274]: Will follow this change with an isolated change doing this across the entire code base https://github.com/functionless/functionless/issues/177

File: src/expression.ts (hunk 3 above)
Comment (sam-goodwin) [reply src/expression.ts:325]: Left over from an experiment. Removed.

File: src/asl.ts [hunk 8]
```diff
@@ -981,14 +985,18 @@ export class ASL {
           InputPath: ASL.toJsonPath(expr.right),
           ResultPath: ASL.toJsonPath(expr.left),
         };
-      } else if (isLiteralExpr(expr.right)) {
+      } else if (
+        isLiteralExpr(expr.right) ||
+        isUnaryExpr(expr.right) ||
+        isBinaryExpr(expr.right)
+      ) {
```
[the remark this answers is not in the document]
Comment (sam-goodwin) [reply src/asl.ts:992]: `anyOf` is useful for creating another type assertion function for re-use elsewhere or for passing an assertion function to another function. I wouldn't use it inline here.

File: test/test-files/step-function.ts (hunk 5 above)
Comment (sam-goodwin) [reply test/test-files/step-function.ts:9]: Anything constantly known should be supported.

**Review by thantos — APPROVED** [review]

**Pull request merged**